vcr/oidc4vci/service/identifiers.go

```diff
@@ -0,0 +1,164 @@
+		return "", fmt.Errorf("unable to resolve %s service: %w", types.BaseURLServiceType, err)
+	}
+	var result string
+	_ = service.UnmarshalServiceEndpoint(&result)
+	if result != "" {
+		result = core.JoinURLPaths(result, "n2n", "identity", url.PathEscape(id.String()))
+	}
+	return result, nil
+}
+
+// NewTLSIdentifierResolver creates a IdentifierResolver that tries to derive the identifier from the TLS certificate if it can't be resolved using the DID document.
+// It does so by constructing the identifier from the CommonName and SubjectAlternativeNames of the certificate and requesting metadata.
+func NewTLSIdentifierResolver(underlying IdentifierResolver, config *tls.Config) IdentifierResolver {
+	result := tlsIdentifierResolver{
+		underlying:       underlying,
+		config:           config,
+		cachedIdentifier: new(atomic.Pointer[string]),
+		lastAttempt:      new(atomic.Pointer[time.Time]),
+	}
+	result.lastAttempt.Store(new(time.Time))
+	return result
+}
+
+const tlsAttemptInterval = time.Minute
+
+var tlsIdentifierResolverPort = 443
+
+func SetTLSIdentifierResolverPort(port int) {
+	tlsIdentifierResolverPort = port
+}
+
+var _ IdentifierResolver = tlsIdentifierResolver{}
+
+type tlsIdentifierResolver struct {
+	underlying       IdentifierResolver
+	config           *tls.Config
+	cachedIdentifier *atomic.Pointer[string]
+	// lastAttempt is the time at which the last attempt to resolve the identifier from the TLS certificate was made.
+	// It is used to prevent spamming the local node, since it could be called on each OpenID4VCI request.
+	lastAttempt *atomic.Pointer[time.Time]
+}
+
+func (t tlsIdentifierResolver) Resolve(id did.DID) (string, error) {
+	identifier, err := t.underlying.Resolve(id)
+	if err != nil {
+		return "", err
+	}
+	if identifier != "" {
+		return identifier, nil
+	}
+	// Could not load from DID document, try to derive from TLS certificate
+	cached := t.cachedIdentifier.Load()
+	if cached != nil {
+		return *cached, nil
+	}
+
+	if time.Since(*t.lastAttempt.Load()) > tlsAttemptInterval {
+		lastAttempt := time.Now()
+		t.lastAttempt.Store(&lastAttempt)
+		identifier, err = t.resolveFromCertificate(id)
+		if err == nil {
+			t.cachedIdentifier.Store(&identifier)
+		}
+		return identifier, err
+	}
+	return "", nil
+}
+
+func (t tlsIdentifierResolver) resolveFromCertificate(id did.DID) (string, error) {
+	// Construct candidate URLs from TLS certificate SANs
+	var candidateURLs []string
+	// Support legacy TLS certificates with host name in Subject.CommonName as well
+	candidateHosts := append(t.config.Certificates[0].Leaf.DNSNames, t.config.Certificates[0].Leaf.Subject.CommonName)
+	for _, host := range candidateHosts {
+		candidateURLs = append(candidateURLs, fmt.Sprintf("https://%s:%d", host, tlsIdentifierResolverPort))
+	}
+
+	// Resolve URLs
+	httpTransport := http.DefaultTransport.(*http.Transport).Clone()
+	httpTransport.TLSClientConfig = t.config
+	httpClient := &http.Client{
+		Timeout:   2 * time.Second,
```

Comment:
extremely short, if you miss a single `SYN` package, timeout is triggered. Recommend 5